lib/timeline/frame_flame_chart.dart

```diff
@@ -32,90 +34,111 @@ class FrameFlameChart extends CoreElement {
-    this.data = data;
+  void updateFrameData(TimelineFrame frame) {
+    this.frame = frame;
 
     content.clear();
 
-    if (_debugEventTrace && data != null) {
+    // TODO(kenzie): Sometimes we see a dump of the event trace that does not
+    //  match what we draw in the flame chart. Fix this.
+    if (_debugEventTrace && frame != null) {
       final StringBuffer buf = new StringBuffer();
-      for (TimelineThread thread in data.threads) {
-        buf.writeln('${thread.name}:');
-        for (TimelineThreadEvent event in data.events) {
-          if (event.threadId == thread.threadId) {
-            event.format(buf, '  ');
-          }
-        }
+      buf.writeln('CPU:');
+      for (TimelineEvent event in frame.cpuEvents) {
+        event.format(buf, '  ');
+      }
+      buf.writeln('GPU:');
+      for (TimelineEvent event in frame.gpuEvents) {
+        event.format(buf, '  ');
       }
       print(buf.toString());
     }
 
-    if (data != null) {
-      _render(data);
+    if (frame != null) {
+      _render(frame);
     }
   }
 
-  void _render(TimelineFrameData data) {
-    const int leftIndent = 130;
+  void _render(TimelineFrame frame) {
+    const int leftIndent = 60;
     const int rowHeight = 25;
 
+    // TODO(kenzie): re-write this scale logic.
     const double microsPerFrame = 1000 * 1000 / 60.0;
     const double pxPerMicro = microsPerFrame / 1000.0;
 
     int row = 0;
 
-    final int microsAdjust = data.frame.startMicros;
+    final int microsAdjust = frame.startTime;
 
     int maxRow = 0;
 
-    void drawRecursively(TimelineThreadEvent event, int row) {
-      if (!event.wellFormed) {
-        print('event not well formed: $event');
-        return;
-      }
-
-      final double start = (event.startMicros - microsAdjust) / pxPerMicro;
+    void drawRecursively(TimelineEvent event, int row) {
+      final double start = (event.startTime - microsAdjust) / pxPerMicro;
       final double end =
-          (event.startMicros - microsAdjust + event.durationMicros) /
-              pxPerMicro;
+          (event.startTime - microsAdjust + event.duration) / pxPerMicro;
 
-      _createPosition(event.name, leftIndent + start.round(),
-          (end - start).round(), row * rowHeight);
+      _drawFlameChartItem(
+        event,
+        leftIndent + start.round(),
+        (end - start).round(),
+        row * rowHeight,
+      );
 
       if (row > maxRow) {
         maxRow = row;
       }
 
-      for (TimelineThreadEvent child in event.children) {
+      for (TimelineEvent child in event.children) {
         drawRecursively(child, row + 1);
       }
     }
 
-    // TODO: investigate if this try/catch is necessary.
-    try {
-      for (TimelineThread thread in data.threads) {
-        _createPosition(thread.name, 0, null, row * rowHeight);
+    void drawCpuEvents() {
+      final CoreElement sectionTitle = div(text: 'CPU', c: 'timeline-title');
+      sectionTitle.element.style.left = '0px';
```

Comment:
nit: I'm not a fan but css style is to omit the unit when the value is zero so write
= '0'; instead of '0px'